zio-blocks/src/main/scala/zio/blocks/schema/Optic.scala

```diff
@@ -202,68 +264,56 @@ sealed trait Prism[F[_, _], S, A] extends Optic[F, S, A] {
 }
 
 object Prism {
-  type Bound[S, A] = Prism[Binding, S, A]
-
-  def apply[F[_, _], S, A <: S](parent: Reflect.Variant[F, S], child: Term[F, S, A]): Prism[F, S, A] =
-    new Case(parent, child)
-
-  def apply[F[_, _], S, T, A](first: Prism[F, S, T], second: Prism[F, T, A]): Prism[F, S, A] =
-    new PrismPrism(first, second)
+  type Bound[S, A <: S] = Prism[Binding, S, A]
 
-  private case class Case[F[_, _], S, A <: S](parent: Reflect.Variant[F, S], child: Term[F, S, A])
-      extends Prism[F, S, A]
-      with Leaf[F, S, A] {
+  def apply[F[_, _], S, A <: S](parent: Reflect.Variant[F, S], child: Term[F, S, A]): Prism[F, S, A] = {
     require((parent ne null) && (child ne null))
+    new PrismImpl(parent, parent, child)
+  }
 
-    private var matcher: Matcher[A] = null
+  def apply[F[_, _], S, T <: S, A <: T](first: Prism[F, S, T], second: Prism[F, T, A]): Prism[F, S, A] = {
+    val u1 = first.asInstanceOf[PrismImpl[F, _, _, _]]
+    val u2 = second.asInstanceOf[PrismImpl[F, _, _, _]]
+    new PrismImpl(
+      u1.structure.asInstanceOf[Reflect.Variant[F, S]],
+      u2.parent.asInstanceOf[Reflect.Variant[F, T]],
+      u2.child.asInstanceOf[Term[F, T, A]]
+    )
+  }
 
-    private def init(F: HasBinding[F]): Unit =
-      if (matcher eq null) {
-        val matchers = F.matchers(parent.variantBinding)
-        matcher = matchers(parent.cases.indexWhere(_.name == child.name)).asInstanceOf[Matcher[A]]
-      }
+  private case class PrismImpl[F[_, _], S, T <: S, A <: T](
```

Comment:
Having multiple classes will ensure megamorphic dispatch to all methods in `Optic` (though I realize there aren't many of those). We should strive to have a single class which implements all methods, making sure the method names don't overlap as necessary.